ingest/workflow/snakemake_rules/split_serotypes.smk

```diff
@@ -0,0 +1,43 @@
+"""
+This part of the workflow handles splitting the data by serotype either based on the 
+NCBI metadata or Nextclade dataset. Could use both if necessary to cross-validate.
+
+    metadata = "data/metadata.tsv"
+    sequences = "data/sequences.fasta"
+
+This will produce output files as
+
+    metadata_{serotype} = "results/metadata_{serotype}.tsv"
+    sequences_{serotype} = "results/sequences_{serotype}.fasta"
+
+Parameters are expected to be defined in `config.transform`.
+"""
+
+rule split_by_ncbi_serotype:
+  """
+  Split the data by serotype based on the NCBI metadata.
+  """
+  input:
+    metadata = "data/metadata.tsv",
+    sequences = "data/sequences.fasta"
+  output:
+    metadata = "results/metadata_{serotype}.tsv",
+    sequences = "results/sequences_{serotype}.fasta"
+  params:
+    ncbi_serotype = lambda wildcards: wildcards.serotype,
+    id_field = config["transform"]["id_field"]
+  shell:
+    """
+    if [[ "{params.ncbi_serotype}" == "all" ]]; then
+      cp {input.metadata} {output.metadata}
+      cp {input.sequences} {output.sequences}
+    else
+      augur filter \
+        --sequences {input.sequences} \
+        --metadata {input.metadata} \
+        --metadata-id-columns {params.id_field} \
+        --query "ncbi_serotype=='{params.ncbi_serotype}'" \
+        --output-sequences {output.sequences} \
+        --output-metadata {output.metadata}
+    fi
```

Comment:
One common pattern that I use to avoid using shell if/else statements inside a Snakemake rule is to split out the logical blocks into their own rules and let Snakemake's own filename pattern matching handle the conditional logic more declaratively. For example, this rule can be split into two separate rules like:

```python
rule split_by_ncbi_serotype:
  """
  Split the data by serotype based on the NCBI metadata.
  """
  input:
    metadata = "data/metadata.tsv",
    sequences = "data/sequences.fasta"
  output:
    metadata = "results/metadata_{serotype}.tsv",
    sequences = "results/sequences_{serotype}.fasta"
  params:
    ncbi_serotype = lambda wildcards: wildcards.serotype,
    id_field = config["transform"]["id_field"]
  shell:
    """
    augur filter \
        --sequences {input.sequences} \
        --metadata {input.metadata} \
        --metadata-id-columns {params.id_field} \
        --query "ncbi_serotype=='{params.ncbi_serotype}'" \
        --output-sequences {output.sequences} \
        --output-metadata {output.metadata}
    """

rule use_all_ncbi_serotypes:
  """
  Use data from all serotypes.
  """
  input:
    metadata = "data/metadata.tsv",
    sequences = "data/sequences.fasta"
  output:
    metadata = "results/metadata_all.tsv",
    sequences = "results/sequences_all.fasta"
  shell:
    """
    cp {input.metadata} {output.metadata}
    cp {input.sequences} {output.sequences}
    """
```

The benefit of this approach is that you keep a one-to-one mapping between rules and shell commands, you get clearer output from Snakemake on the terminal (see example below), and also Snakemake won't try to re-run the `use_all_ncbi_serotypes` rule when the `params` in the `split_by_ncbi_serotype` rule change which is [the kind of thing Snakemake does nowadays](https://snakemake.readthedocs.io/en/stable/project_info/history.html#id206).

An example of the printed shell commands from a Snakemake dryrun with the current PR state is:

```
    if [[ "denv4" == "all" ]]; then
      cp data/metadata.tsv results/metadata_denv4.tsv
      cp data/sequences.fasta results/sequences_denv4.fasta
    else
      augur filter         --sequences data/sequences.fasta         --metadata data/metadata.tsv         --metadata-id-columns genbank_accession         --query "ncbi_serotype=='denv4'"         --output-sequences results/sequences_denv4.fasta         --output-metadata results/metadata_denv4.tsv
    fi
```

vs.

```
    augur filter     --sequences data/sequences.fasta     --metadata data/metadata.tsv     --metadata-id-columns genbank_accession     --query "ncbi_serotype=='denv4'"     --output-sequences results/sequences_denv4.fasta     --output-metadata results/metadata_denv4.tsv
```

Reply:
This deserves an entry in the [Nextstrain Snakemake style guide](https://docs.nextstrain.org/en/latest/reference/snakemake-style-guide.html) 🌟

Reply:
The if/then statement was dropped via a refactor and simplification in https://github.com/nextstrain/dengue/commit/0537e23053972cdee9f2514769ef572143032477 Thanks!